hail/src/main/scala/is/hail/expr/types/physical/PType.scala

```diff
@@ -304,4 +305,47 @@ abstract class PType extends BaseType with Serializable with Requiredness {
   def unify(concrete: PType): Boolean = {
     this.isOfType(concrete)
   }
+
+  // Semantics: must be callable without requiredeness check: srcAddress must point to non-null value
+  def copyFromType(mb: MethodBuilder, region: Code[Region], srcPType: PType, srcAddress: Code[Long], forceDeep: Boolean): Code[Long] = {
+    this.fundamentalType match {
+      case _: PBoolean | _: PInt32 | _: PInt64 | _: PFloat32 | _: PFloat64 =>
+        srcAddress
+      case ft => throw new UnsupportedOperationException("Unknown fundamental type: " + ft)
+    }
+  }
+  def copyFromType(mb: MethodBuilder, region: Code[Region], srcPType: PType, srcAddress: Code[Long]): Code[Long] =
+    this.copyFromType(mb, region, srcPType, srcAddress, false)
+
+  def copyFromType(region: Region, srcPType: PType, srcAddress: Long, forceDeep: Boolean): Long = {
+    this.fundamentalType match {
+      case _: PBoolean | _: PInt32 | _: PInt64 | _: PFloat32 | _: PFloat64 =>
+        srcAddress
+      case ft => throw new UnsupportedOperationException("Unknown fundamental type: " + ft)
+    }
+  }
+  def copyFromType(region: Region, srcPType: PType, srcAddress: Long): Long =
+    this.copyFromType(region, srcPType, srcAddress, false)
+
+  def storeShallowAtOffset(dstAddress: Code[Long], srcAddress: Code[Long]): Code[Unit] = {
```

Comment:
why is this implemented on PType and not on these types?

Reply:
Just to keep the implementation less verbose, seemed to be something that PType was used for (similar to `setRequired`, `isNumeric`, `canonical`)